back/engines/commercial/multi_tenancy/lib/tasks/core/translate_project_to_all_locales.rake

```diff
@@ -0,0 +1,117 @@
+  # It also assumes a user with email 'moderator@citizenlab.co' or 'moderator@govocal.com' exists.
+  #
+  # It will overwrite any existing translations.
+  desc 'Translate project multilocs to all tenant locales'
+  task :translate_project_to_tenant_locales, %i[host project_id] => :environment do |_t, args|
+    Apartment::Tenant.switch(args[:host].tr('.', '_')) do
+      project = Project.find(args[:project_id])
+      locales = AppConfiguration.instance.settings['core']['locales']
+      translate_project = TranslateProjectMethods.new
+
+      puts "\nTranslating project: '#{project.title_multiloc['en']}' to all tenant locales\n\n"
+
+      translate_project.process_model(project, locales)
+
+      project.phases.each do |phase|
+        translate_project.process_model(phase, locales)
+
+        fields = phase&.custom_form&.custom_fields
+        fields&.each do |field|
+          translate_project.process_model(field, locales)
+          options = field&.options
+          options&.each { |option| translate_project.process_model(option, locales) }
+        end
+
+        report = phase&.report
+        layout = report&.layout
+        translate_project.process_layout(layout, locales) if layout
+        translate_project.process_data_units(report) if report
+      end
+
+      puts "\n--- end ---\n\n"
+    end
+  end
+end
+
+class TranslateProjectMethods
+  def process_model(model, locales)
+    multiloc_attributes = model.attributes.select { |attribute| attribute.ends_with? '_multiloc' }
+
+    multiloc_attributes.each do |multiloc_attribute|
+      attribute_name = multiloc_attribute[0]
+      translated = process_multiloc(multiloc_attribute[1], locales)
+      next unless translated
+
+      model[attribute_name] = translated
+    end
+
+    save_model(model)
+  end
+
+  def process_layout(layout, locales)
+    craftjs_json = layout.craftjs_json
+
+    craftjs_json.each do |k, v|
+      next unless v.is_a?(Hash)
+
+      text = v.dig('props', 'text')
+      next unless text.is_a?(Hash) && text['en'].present?
+
+      craftjs_json[k]['props']['text'] = process_multiloc(text, locales)
+    end
+
+    layout.craftjs_json = craftjs_json
+
+    save_model(layout)
+  end
+
+  def process_data_units(report)
+    user = User.find_by(email: 'moderator@citizenlab.co') || User.find_by(email: 'moderator@govocal.com')
+
+    if user.blank?
+      puts "No moderator user found, so could not re-publish graph_data_units for Report: #{report.id}"
+      nil
+    end
+
+    ReportBuilder::ReportPublisher.new(report, user).publish
+
+    puts "Re-published graph_data_units for Report: #{report.id}"
+  end
+
+  def process_multiloc(multiloc, locales)
+    english = multiloc['en']
+    return nil unless english
+
+    translator = MachineTranslations::MachineTranslationService.new
+
+    locales.each do |locale|
+      next if locale == 'en'
+
+      multiloc[locale] = translator.translate(english, 'en', locale)
```

Comment:
If there are two locales with the same language (for example "fr-FR" and "fr-BE"), we would translate twice, which takes more time and is more expensive. Perhaps we could consider if there already exists a translation for a locale of the same language in the `multiloc` and consider reusing that one first? 

If there are not multiple locales of the same language for the platform you're planning to use it with, then you don't have to spend the additional effort, but perhaps it could be useful to add a comment to explain this.

Reply:
Good idea. In this case there aren't any such locale collisions, but it's a good thing to note if we develop the task further.